(obsolete - do not merge) [Inference API] Add Azure AI Studio Embeddings and Chat Completion Support

markjhoy opened the pull request:
This PR adds support for Azure AI Studio integration into the Inference API. Currently this supports `text_embedding` and `completion` task types.

### Prerequisites to Model Creation
* You must have an Azure subscription with [Azure AI Studio](https://ai.azure.com/) access
* You must have a [deployed model](https://ai.azure.com/explore/models) either Chat Completion or Embeddings

![image](https://github.com/elastic/elasticsearch/assets/13320310/f8e4d600-c22b-42e8-9f2d-5dd7b1c52ba6)


### Model Creation:
```
PUT _inference/{tasktype}/{model_id}
{
  "service": "azureaistudio",
  "service_settings": {
    "api_key": "{api_key}",
    "target": “{deployment_target}”,
    “provider”: “(model provider}”,
    “endpoint_type”: “(endpoint type)”
  }
}
```
* Valid `{tasktype}` types are: [`text_embedding`, `completion`]

### Required Service Settings:
* `api_key`: The API key can be found on your Azure AI Studio deployment's overview page
* `target`: The target URL can be found on your Azure AI Studio deployment's overview page
* `provider`: Valid provider types are (case insensitive): 
  * `openai` - available for embeddings and completion
  * `mistral` - available for completion only
  * `meta` - available for completion only
  * `microsoft_phi` - available for completion only
  * `cohere` - available for embeddings and completion
  * `snowflake` - available for completion only
  * `databricks` - available for completion only
* `endpoint_type`: Valid endpoint types are:
  * `token` - a "pay as you go" endpoint (charged by token)
    * Available for OpenAI, Meta and Cohere
  * `realtime` - a realtime endpoint VM deployment (charged by the hour)
    * Available for Mistral, Meta, Microsoft Phi, Snowflake and Databricks

### Embeddings Service Settings
* `dimensions`: (optional) the number of dimensions the resulting output embeddings should have. 

### Embeddings Task Settings
(this is also overridable in the inference request)
* `user`: (optional) a string that is a unique identifier representing your end-user. This helps Azure AI Studio in the case of abuse or issues for debugging.

### Completion Service Settings
(no additional service settings)

### Completion Task Settings
(these are all optional and can be overridden in the inference request)
* `temperature`: What sampling temperature to use, between 0 and 2. Higher values mean the model takes more risks. Try 0.9 for more creative applications, and 0 (argmax sampling) for ones with a well-defined answer. Microsoft recommends altering this or top_p but not both.
* `top_p`: An alternative to sampling with temperature, called nucleus sampling, where the model considers the results of the tokens with top_p probability mass. So 0.1 means only the tokens comprising the top 10% probability mass are considered. Microsoft recommends altering this or temperature but not both.
* `do_sample`: request to perform the sampling or not
* `max_new_tokens`: the maximum number of new tokens the chat completion inference should produce in the output

### Text Embedding Inference
```
POST _inference/text_embedding/{model_id}
{
    "input": "The answer to the universe is"
}
```

### Chat Completion Inference
```
POST _inference/completion/{model_id}
{
    "input": "The answer to the universe is"
}
```

Commit:
initial implementation azure ai studio _inference

Commit:
Add embeddings service settings model tests

Commit:
add completion service settings tests

Commit:
adds chat completion and embeddings request tests

Commit:
add tests for completion and embeddings responses

Commit:
add action and action creator test coverage

Commit:
remove forbidden API call to .getBytes()

Commit:
rename "Completions" to "ChatCompletions"

Commit:
adding service tests

shainaraskas committed:
[DOCS] Add API example + diagrams to shard allocation awareness docs (#108390)

nik9000 committed:
ESQL: Move a few more test out of IT_test_only (#108377)

This moves the "skip" logic from our IT_test_only suffix into a new
feature - this one is historical `esql.enrich_load`. This feature is not
supported by `CsvTests` but is supported across all tests.

DiannaHohensee committed:
Brief document blurb about RestClient (#107863)

Commit:
nativeaccess: try to load all located libsystemds (#108238)

Linux systems with multiarch (e.g. i386 & x86_64) libraries
may have libsystemd.0 in two subdirectories of an entry in
java.library.path. For example, libsystemd.so.0 may be found
in both /usr/lib/i386-linux-gnu and /usr/lib/x86_64-linux-gnu.

Instead of attempting to load any library found, attempt all
and stop as soon as one is successfully loaded.

dnhatn committed:
Exchange should wait for remote sinks (#108337)

Today, we do not wait for remote sinks to stop before completing the 
main request. While this doesn't affect correctness, it's important  that
we do not spawn child requests after the parent request is completed.

Closes #105859

Commit:
finalize service settings test

dakrone committed:
Move kibana reporting data stream settings into component template (#107581)

Previously these were contained in the index template, however, Kibana
needs to be able to make overrides to only the settings, so factoring
these out would allow them to do this (in such a way that they can be
overridden by the `kibana-reporting@custom` component template as well).

Relates to #97765

rjernst committed:
Mock empty state task in file settings tests (#108100)

When the file watched by file settings is initially missing, a special
method in reserved state service is called to write a dummy cluster
state entry. In the case of tests, there is no real running master
service, so when the task is submitted, the file watcher thread actually
barfs and the watcher dies, silently. That then causes the test to
timeout as it waits indefinitely but the file watcher is no longer
watching for the test file that was written.

This commit mocks out writing this empty state in the reserved state
service. It also collapses the two tests that check stopping while
blocked in processing works since they were almost exactly the same.

closes #106968

dakrone committed:
Mention alias filters don't apply for get-by-id in docs (#108433)

Resolves #3861

rjernst committed:
Make MockLogAppender threadsafe (#108206)

Adding and removing appenders in Log4j is not threadsafe. Yet some tests
rely on capturing logging by adding an in memory appender,
MockLogAppender. This commit makes the mock logging threadsafe by
creating a new, singular appender for mock logging that delegates, in a
threadsafe way, to the existing appenders created. Confusingly
MockLogAppender is no longer really an appender, but I'm leaving
clarifying that for a followup so as to limit the scope of this PR.

closes #106425

costin committed:
ESQL: Disable quoting in FROM command (#108431)

Disable location quoting in FROM command before 8.14 release to allow
 more time to discuss options

dnhatn committed:
Add factories for time series aggregation (#107803)

This change introduces operator factories for time-series aggregations. 
A time-series aggregation executes in three stages, deviating from the
typical two-stage aggregation.

For example: `sum(rate(write_requests)), avg(cpu) BY cluster, time-bucket`

**1. Initial Stage:** In this stage, a standard hash aggregation is
executed, grouped by tsid and time-bucket. The `values` aggregations are
added to collect values of the grouping keys excluding the time-bucket,
which are then used for final result grouping.

```
rate[INITIAL](write_requests), avg[INITIAL](cpu), values[SINGLE](cluster) BY tsid, time-bucket
```

**2. Intermediate Stage:** Equivalent to the final mode of a standard
hash aggregation. This stage merges and reduces the result of the rate
aggregations, but merges without reducing the results of non-rate
aggregations. Certain aggregations, such as count_distinct, cannot have
their final results combined.

```
rate[FINAL](write_requests), avg[INTERMEDIATE](cpu), values[SINGLE](cluster) BY tsid, time-bucket
```

**3. Final Stage:** This extra stage performs outer aggregations over
the rate results and combines the intermediate results of non-rate 
aggregations using the specified user-defined grouping keys.

```
sum[SINGLE](rate_result), avg[FINAL](cpu) BY cluster, bucket
```

luigidellaquila committed:
ES|QL: account for page overhead when calculating memory used by blocks (#108347)

nicktindall committed:
[DOCS] Fix stored_fields parameter description (#98385) (#108445)

(referenced from get and multi_get API docs)

Closes #98385

Commit:
apm-data: ignore malformed fields, and too many dynamic fields (#108444)

* apm-data: ignore_{malformed,dynamic_beyond_limit}

Enable ignore_malformed on all non-metrics APM data streams,
and enable ignore_dynamic_beyond_limit for all APM data streams.

We can enable ignore_malformed on metrics data streams when
https://github.com/elastic/elasticsearch/issues/90007 is fixed.

* Update docs/changelog/108444.yaml

szabosteve committed:
[DOCS] Adds complete Cohere tutorial (#108415)

DaveCTurner committed:
Move conceptual docs about `ActionListener` (#107875)

This information is more discoverable as the class-level javadocs for
`ActionListener` itself rather than hidden away in a separate Markdown
file. Also this way the links all stay up to date.

thecoop committed:
Fix race in SpawnerNoBootstrapTests (#108416)

davidkyle reviewed (commented):
Looks great

It's a very large PR I'd like to do some testing then take another look

markjhoy commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/sender/AzureAiStudioEmbeddingsRequestManager.java:
```diff
@@ -0,0 +1,65 @@
+import static org.elasticsearch.xpack.inference.common.Truncator.truncate;
+
+public class AzureAiStudioEmbeddingsRequestManager extends AzureAiStudioRequestManager {
+    private static final Logger logger = LogManager.getLogger(AzureAiStudioChatCompletionRequestManager.class);
```
Ah - nice catch. The joys of similar base code that's copied and pasted :)

markjhoy commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/ChatCompletionResponseEntity.java:
```diff
@@ -0,0 +1,27 @@
+import java.io.IOException;
+
+/**
+ * TODO - fill in comment
```
Ooof... I did a "Find All" to look for any `TODO` markers I had left... looks like I missed this one ;)

markjhoy commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureaistudio/AzureAiStudioEndpointType.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeEnum(this);
```
Ah - that's a good idea... I'll take a look

markjhoy replied:
> and use the read/writeEnum methods.
🤦  - I realized that in the ServiceSettings superclass, I am actually using these methods, so the `Writable` implementation is not needed on either this or the `Provider` enums...

szabosteve committed:
[DOCS] Fixes typo in Cohere ES tutorial (#108456)

* [DOCS] Fixes typo in Cohere ES tutorial.

* [DOCS] Fixes list.

nik9000 committed:
ESQL: Remove remaining IT_tests_only (#108434)

This moves examples from files marked to run in integration tests only
to the files where they belong and disables this pattern matching. We
now use supported features.

Commit:
cleanups; add missing comments :facepalm:

mayya-sharipova committed:
Correct typo in documentation (#108462)

Correct a small typo: one closing ">" was missing.

kkrik-es committed:
Version-guard checking for lossy params in _source (#108460)

markjhoy commented:
buildkite test this

dnhatn committed:
Log running tasks in EsqlDisruptionIT (#108440)

This PR logs tasks that are running after the disruption is cleared, 
allowing us to investigate why the disruption tests failed in #107347.

Relates #107347

astefan committed:
ESQL: Add aggregates node level reduction (#107876)

* Add aggregation intermediate reduction level and estimatedRowSize
computed value

not-napoleon committed:
Add an APM metric to aggregations usage (#108110)

This wires up the "new" APM metrics integration to the existing Aggregations usage tracking system. It introduces one new metric, a LongCounter named es.search.query.aggregations.total, which has dimensions for the specific aggregation being run, and the values source type we resolved it to.

---------

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

not-napoleon committed:
Fix ClassCastException in Significant Terms (#108429)

Prior to this PR, if a SignificantTerms aggregation targeted a field existing on two indices (that were included in the aggregation) but mapped to different field types, the query would fail at reduce time with a somewhat obscure ClassCastException. This change brings the behavior in line with the Terms aggregation, which returns a 400 class IllegalArgumentException with a useful message in this situation.

Resolves #108427

jonathan-buttner commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureaistudio/AzureAiStudioService.java:
```diff
@@ -0,0 +1,348 @@
+    ) {
+        var actionCreator = new AzureAiStudioActionCreator(getSender(), getServiceComponents());
+
+        if (model instanceof AzureAiStudioEmbeddingsModel embeddingsModel) {
```
I mentioned this above, but the reason I chose the visitor pattern for the other services was to let the model dictate the action that is used. That way we don't need the `instanceof` checks here, except one to ensure that it's a `AzureAiStudioModel` model with an `accept()` method.

Commit:
commit after rebase

markjhoy commented:
Oh - I sometimes hate git and it's ability to not allow a clean merge... going to close this and re-open a new one that's clean.

jonathan-buttner reviewed (commented):
I'm going to add these here so I don't lose them haha, I'll copy them over to the new PR when it's up 👍

markjhoy commented:
Sorry @jonathan-buttner ! :) -- new PR is up here: https://github.com/elastic/elasticsearch/pull/108472
